[algolia-insights] add support for auth user token and product list viewed

jkaho opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

Add support for the Algolia Insights API optional [`authenticatedUserToken`](https://www.algolia.com/doc/api-reference/api-methods/viewed-object-ids/#parameters) field.

Add support for the Segment [`Product List Viewed`](https://segment.com/docs/connections/spec/ecommerce/v2/#product-list-viewed) event.

## Testing

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [If destination is already live] Tested for backward compatibility of destination. **Note:** New required fields are a breaking change.
- [ ] [Segmenters] Tested in the staging environment
- [ ] [Segmenters] [If applicable for this change] Tested for regression with Hadron.

joe-ayoub-segment commented:
Thanks for the PR @jkaho . I'll schedule time to review.

joe-ayoub-segment commented:
Hi @jkaho I had a quick scan. Did you mean to include the file at this location? packages/browser-destinations/destinations/algolia-plugins/src/algoliaInsights/generated-types.ts

joe-ayoub-segment commented:
Hi again @jkaho , I reviewed the PR. 

Regarding adding the new field `authenticatedUserToken`: 

- Looks like this field is to capture an identifier for a user when the user is known to the customer. Does this mean that the `userToken` field was always intended to capture an anonymous user's anonymous identifier? because it used to try and capture the main identifier and only fell back to the anonymous one. 
- How will this change affect customers who are already sending the Segment userId (which is the id for users who are known to the customer) to the `userToken` field? They won't be populating the `authenticatedUserToken` field (so it will be undefined).  
- Are you able to provide evidence of testing this please?

Regarding the new Action: 

- I don't think the default mapping used for `ObjectIds` field will work. Have you tested it locally?
- Are you able to provide evidence of testing this please?

Best regards,
Joe

joe-ayoub-segment commented:
hi again @jkaho can you please also run `yarn types` and commit the updated file please?

jkaho commented:
Hey @joe-ayoub-segment, appreciate the review!

* We added the [`authenticatedUserToken`](https://www.algolia.com/doc/guides/sending-events/concepts/usertoken/#persistent-user-token) field a couple of years ago now, before which the `userToken` field was intended to capture either the auth or anon token. The addition of the new auth field meant we changed the way we recommend users send us user identifiers, though this wasn't a breaking change. If users don't set the auth token field, we will take the value from the user token field (which is required). Again, this change was made a little while ago now and neglecting to add support for it to our Segment connector was an oversight on our part. Is there anything you would recommend in terms of documentation to mitigate potential confusion?
* I have tested the object IDs mapping via the Actions Tester, and it does seem to map the correct values as per the highlighted portions in the screenshot below. Have I misinterpreted something here? Though, I am noticing that the input badge doesn't show in the user configuration UI..

<img width="1912" height="963" alt="Screenshot 2025-08-12 at 10 32 37 am" src="https://github.com/user-attachments/assets/e8dee380-cc6d-437c-ad56-b0bcbd2ef66b" />

joe-ayoub-segment commented:
> Hey @joe-ayoub-segment, appreciate the review!
> 
> * We added the [`authenticatedUserToken`](https://www.algolia.com/doc/guides/sending-events/concepts/usertoken/#persistent-user-token) field a couple of years ago now, before which the `userToken` field was intended to capture either the auth or anon token. The addition of the new auth field meant we changed the way we recommend users send us user identifiers, though this wasn't a breaking change. If users don't set the auth token field, we will take the value from the user token field (which is required). Again, this change was made a little while ago now and neglecting to add support for it to our Segment connector was an oversight on our part. Is there anything you would recommend in terms of documentation to mitigate potential confusion?
> * I have tested the object IDs mapping via the Actions Tester, and it does seem to map the correct values as per the highlighted portions in the screenshot below. Have I misinterpreted something here? Though, I am noticing that the input badge doesn't show in the user configuration UI..
> 
> <img alt="Screenshot 2025-08-12 at 10 32 37 am" width="1912" height="963" src="https://private-user-images.githubusercontent.com/72959522/476888992-e8dee380-cc6d-437c-ad56-b0bcbd2ef66b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUwODA1NDUsIm5iZiI6MTc1NTA4MDI0NSwicGF0aCI6Ii83Mjk1OTUyMi80NzY4ODg5OTItZThkZWUzODAtY2M2ZC00MzdjLWFkNTYtYjBiY2JkMmVmNjZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODEzVDEwMTcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkODE3ZGQwZWEyNmZiMDM0MzAzNDIyYjczN2ZkODk2MjEyNDU3ODg1MzgzNjg4ZTZjZjUxMzY2MWRhYThmYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ij2CetQGooTtJsaPLOu8tb1d-v71Wjh3uDc4XUOWFS4">

Hi @jkaho , that's correct, the 'badge' contains the mapping details the user gets to see in the UI. Our UI doesn't support this type of mapping, so while the mapping itself might work, it breaks the UI. 

I suggest changing the mapping to something like this

```
{
        '@path': '$.properties.products.product_ids'
}
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/algolia-insights/productListViewedEvents/index.ts:
```diff
@@ -0,0 +1,144 @@
+      multiple: true,
+      required: true,
+      properties: {
+        product_id: { label: 'product_id', type: 'string', required: true }
```
Please add a description for product_id.

joe-ayoub-segment replied:
If you want to prevent the customer from adding additional values into items in the product array, you can add this: 

  `additionalProperties: false,`

jkaho replied:
Ah awesome, thanks for the feedback 🙂

joe-ayoub-segment commented:
Thanks @jkaho this looks good. Only a nit issue left. Please see the comments. 
Best regards,
Joe

joe-ayoub-segment commented:
Thanks @jkaho I've labelled it for deployment. I'll notify you once it's out. It might be up to 7 days, as we usually only deploy this repo weekly and we missed this week's deployment.

jkaho commented:
@joe-ayoub-segment is there anything for me to action with the failing CI?

joe-ayoub-segment commented:
Hi @jkaho yes looks like snapshot tests are failing. Can you run this command and commit the changes please?

`yarn types `


FAIL  src/destinations/algolia-insights/productListViewedEvents/__tests__/snapshot.test.ts
    ● Testing snapshot for AlgoliaInsights's productListViewedEvents destination action: › required fields
  
      expect(received).toMatchSnapshot()
  
      Snapshot name: `Testing snapshot for AlgoliaInsights's productListViewedEvents destination action: required fields 1`
  
      - Snapshot  - 0
      + Received  + 5
  
      @@ -6,10 +6,15 @@
              "eventType": "view",
              "index": "lx7Lx",
              "objectIDs": Array [
                "lx7Lx",
              ],
      +       "products": Array [
      +         Object {
      +           "product_id": "lx7Lx",
      +         },
      +       ],
              "timestamp": 1674843786677,
              "userToken": "lx7Lx",
            },
          ],
        }

jkaho commented:
@joe-ayoub-segment if you're around, it would be awesome if we were able to get this out in the next deployment (which I believe is tomorrow?) 🙏
